### PR TITLE
Support panic and trap for jvm target 

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
Support panic and trap statement in bir and jvm codegen.

## Approach
This section covers the error type and value, panic and trap constructs. Here error handling will be described in two sub sections with BIR and JVM target.

### BIR Design
**Error type and literal constructor**

```
%8 = new_error %9 %10;
```
**Panic**
Panic will be a terminator in BIR.

```
%4 = new_error %5 %6;
panic %4;
```

**Trap**
We will use error table to mark the trap expression in BIR. New basic block will be created for each trap expression. 
```
	Error Table 
		BB	|errorOp
		bb1	|%6
		bb3	|%11
		bb4	|%10
```

### JVM Code Gen
**Ballerina construct** | **JVM Representation**
-- | --
error | Runtime Exception
panic | Throw
trap | Try catch

## Samples

**Panic example** 
ex. 
```
type AccountNotFoundError error<string, string>;
function getAccountBalance(int accountID) returns int {
      
        string errorReason = "Account Not Found";
        AccountNotFoundError accountNotFoundError =  error(errorReason,"3");
        if (accountID == 0){
            panic accountNotFoundError;
        }
        return accountID;
    
}

public function main() returns int {
    var result = getAccountBalance(0);
	return result;
}

```
this will result in 

Exception in thread "main" Account Not Found 3
	at hello.getAccountBalance(Unknown Source)
	at hello.main(Unknown Source)
	at hello.main(Unknown Source)

**Trap sample** 
```
public function main() returns error? {
   var x =   trap foo(0);
   if ( x is error) {
       return x;
   }
}
public function foo(int x) returns error|int  {
     if(x == 0){
        panic error("reason", {"message":"detail"});
    } else {
        return x;
    }
}
```

this will result in

reason {"message":"detail"}

## Remarks
We use java stacktrace directly when printing. This required to be formatted to comply with bal stacktrace.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by Kishanthan** [comment]
@warunalakshitha , conflicts, lets fix them

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/BuilderUtils.java [hunk 1]
```diff
@@ -146,13 +146,17 @@ public static void compileAndWriteJar(Path sourceRootPath,
                                           boolean offline,
                                           boolean lockEnabled,
                                           boolean skipTests,
-                                          boolean enableExperimentalFeatures) {
+                                          boolean enableExperimentalFeatures, 
+                                          boolean dumpBIR) {
```
Comment (vinok88) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/BuilderUtils.java]: Did we decide to add this?

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_method_gen.bal [hunk 2]
```diff
@@ -26,6 +26,7 @@ function generateMethod(bir:Function func, jvm:ClassWriter cw, bir:Package modul
     // generate method desc
     string desc = getMethodDesc(func);
     jvm:MethodVisitor mv = cw.visitMethod(ACC_PUBLIC + ACC_STATIC, funcName, desc, (), ());
+    InstructionGenerator instGen = new(mv, indexMap, currentPackageName);
```
Comment (vinok88) [inline compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_method_gen.bal]: Why did we need this change?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/BuilderUtils.java (hunk 1 above)
Comment (warunalakshitha) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/BuilderUtils.java]: Yes

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_method_gen.bal (hunk 2 above)
Comment (warunalakshitha) [reply compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_method_gen.bal]: Previously InstructionGenerator was created for each instruction (inside the loop) . Hence I moved out of the loop so we can reuse same object.

**Pull request merged**